Include FAR in Medik8s

razo7 opened the pull request:
- Update the link and name for FAR on the remediation page
- Add five FAR pages- about, configuration, how it works, known issues, and limitations pages

razo7 committed:
Include FAR as alternative remeidator in remediation page

Update link and name for FAR in remediation page

razo7 committed:
Add FAR pages

Add about, configuration, how it works, known issues, and limitations pages

clobrano commented on remediation/fence-agents-remediation/Limitations.md:
```diff
@@ -0,0 +1,12 @@
+
+## Limitations
+
+1. FAR is limited to running FA with only one [action](https://github.com/ClusterLabs/fence-agents/blob/main/doc/FenceAgentAPI.md#agent-operations-and-return-values), reboot.
```
- It is preferable to use the full name "Fence Agents" first before using an acronym. It's kind of annoying to do in each page, but you'll never know which first page a reader will face (maybe they just arrived here from some external link).
- Remediation time can be long, not slow.

razo7 replied:
Good points

clobrano commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,27 @@
+
+[Fence Agents Remediation](https://github.com/medik8s/fence-agents-remediation#readme) (FAR) is a Kubernetes operator that uses external tools to remediate unhealthy nodes. The tools are a set of well-known [upstream fence agents (FA)](https://github.com/ClusterLabs/fence-agents), where each one of them can be used for different environments, to fence a node using a reboot action. Moreover, FAR c ...
+
+FAR is recommended when a node becomes unhealthy, and we want to completely isolate the node from a cluster, since we can’t “trust” the unhealthy node, to prevent it from accessing the shared resources like [RWO volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes). The operator is available in the Kuberenets community, [OperatorHub.io](https://operatorhub.io/oper ...
```
two typos
- s/Kuberenets/Kubernetes
- s/docuementaion/documentation

clobrano commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,27 @@
+* Robustness - FAR can remediate an unhealthy node using a management interface (e.g., IPMI) while still keeping control plane connectivity.
+* Speed - FAR is fast as it adds the FAR NoExecute taint, and deletes workloads to accelerate workload rescheduling, in addition to rebooting the node using a fence agent.
+* Diversity - FAR includes several FAs from a whole known set of upstream fencing agents for bare metal servers, virtual machines, cloud platforms, etc.
+* Adjustability - FAR allows to set up of different parameters to access the management interface that remediates the node.
```
> set up of different

"of" is not needed here.

mshitrit commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -0,0 +1,17 @@
+
+* Install FAR and NHC
+* Create FAR Template CR
+* Create NHC CR to use FAR Template CR accordingly (see [FAR documentation](https://github.com/medik8s/fence-agents-remediation#far-with-nhc)).
```
Another option is to use NHC's escalation remediation

razo7 replied:
>Create NHC CR to use FAR Template CR accordingly

But writing the above includes NHC's escalation remediation as a user can create NHC CR to solely use FAR template or FAR template with different remediation templates.. 
Do you still see a point of pointing this out here? I might add this in https://github.com/medik8s/fence-agents-remediation#far-with-nhc if it isn't included (after the new FAR readme PR) or in NHC documentation.

slintes replied:
both classic as well as escalating remediation use the template, so IMHO this is fine

mshitrit commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,27 @@
+Generally Available
+{: .label .label-green }
+
+[Fence Agents Remediation](https://github.com/medik8s/fence-agents-remediation#readme) (FAR) is a Kubernetes operator that uses external tools to remediate unhealthy nodes. The tools are a set of well-known [upstream fence agents (FA)](https://github.com/ClusterLabs/fence-agents), where each one of them can be used for different environments, to fence a node using a reboot action. Moreover, FAR c ...
```
I think the terminology here is confusing, It looks like the reboot done by the FA is synonymous to remediation.
I think this terminology is wrong, the reboot IMO would be the fencing where the remediation is the complete process which includes the workload deletion and Node recovery.

mshitrit commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,27 @@
+### Advantages
+
+* Robustness - FAR can remediate an unhealthy node using a management interface (e.g., IPMI) while still keeping control plane connectivity.
+* Speed - FAR is fast as it adds the FAR NoExecute taint, and deletes workloads to accelerate workload rescheduling, in addition to rebooting the node using a fence agent.
```
Workload deletion is critical and not just meant to accelerate (at least as far as statefull workload)

razo7 replied:
I added a paragraph with a comment about that. Let me know if that's good enough for expressing that or if should I rephrase the sentence.

clobrano commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,18 @@
+Fence Agents Remediation (*FAR*) operator uses a management interface or a traditional API to remediate unhealthy kubernetes nodes.
+It was tested with [Node HealthCheck Operator](https://github.com/medik8s/node-healthcheck-operator) (NHC), that creates a Custom Resource (CR) for FAR with information about the unhealthy Node.
+
+FAR remediates a node by rebooting an unhealthy node, and moving any remaining workloads from the node to other nodes, so they can be procced in other nodes and be isolated from the unhealthy node. The reboot is done by executing a fence agent for the unhealthy node, while evicting the workloads from this node is achieved by tainting the node and deleting the workloads.
```
typo s/proceed/processed?

razo7 committed:
Typo and fencing vs remediation separation

Correcting code review comments

slintes commented on remediation/fence-agents-remediation/known-issues.md:
```diff
@@ -0,0 +1,11 @@
+
+## Known Issues
+
+1. FAR can't be updated from v0.1.0 to any newer versions, since there was a change in the commonLabels to one of [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/) in FAR v0.2.0.
```
The label isn't the reason. The reason is us not dealing with the known issue...
I would skip the reason at all 🤷🏼‍♂️

razo7 replied:
SGTM the reason is not so important information for a TP release.

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,18 @@
+nav_order: 1
+---
+
+## How Fence Agents Remediation works?
```
either remove the question mark, or change to "How does ... work?"

slintes commented on remediation/remediation.md:
```diff
@@ -49,11 +51,12 @@ remediation.  For this reason, the [ExternalRemediation API](https://github.com/
 * [(wip) metal3](https://github.com/metal3-io/cluster-api-provider-metal3/pull/157) - a mechanism designed for bare metal clusters with a functioning Metal3 API
 * [machine-deletion-remediation](https://github.com/medik8s/machine-deletion-remediation) - a mechanism designed for any cluster with a functioning Machine API
-* [(planned) direct]() - a mechanism designed around an existing set of [upstream fencing agents](https://github.com/ClusterLabs/fence-agents) for environments with a traditional API end-point (eg. IPMI) for power cycling cluster nodes
+* [fence-agents-remediation](https://github.com/medik8s/fence-agents-remediation) - a mechanism designed around an existing set of [upstream fencing agents](https://github.com/ClusterLabs/fence-agents) for environments with a traditional API end-point or a management interface (eg. IPMI) for power-cycling the cluster nodes
```
IPMI is providing an API, so why the change from "API end-point (eg. IPMI)" to "API end-point or a management interface (eg. IPMI)"?

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,18 @@
+
+## How Fence Agents Remediation works?
+
+Fence Agents Remediation (*FAR*) operator uses a management interface or a traditional API to remediate unhealthy kubernetes nodes.
```
what's the difference between management interface and traditional API?

razo7 replied:
For this context there is no much difference, and using `traditional API` or just ` API` should be enough to explain how the fence agent is communicating.

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,18 @@
+## How Fence Agents Remediation works?
+
+Fence Agents Remediation (*FAR*) operator uses a management interface or a traditional API to remediate unhealthy kubernetes nodes.
+It was tested with [Node HealthCheck Operator](https://github.com/medik8s/node-healthcheck-operator) (NHC), that creates a Custom Resource (CR) for FAR with information about the unhealthy Node.
```
"It was tested with" -> "It is designed to be used with"?

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,18 @@
+Fence Agents Remediation (*FAR*) operator uses a management interface or a traditional API to remediate unhealthy kubernetes nodes.
+It was tested with [Node HealthCheck Operator](https://github.com/medik8s/node-healthcheck-operator) (NHC), that creates a Custom Resource (CR) for FAR with information about the unhealthy Node.
+
+FAR remediates a node by rebooting an unhealthy node, and moving any remaining workloads from the node to other nodes, so they can be processed in other nodes and be isolated from the unhealthy node. The reboot is done by executing a fence agent for the unhealthy node, while evicting the workloads from this node is achieved by tainting the node and deleting the workloads.
```
"and moving any remaining workloads from the node to other nodes, so they can be processed in other nodes and be isolated from the unhealthy node."

maybe a bit less "nodes" are good enough?
-> "and evicting workloads from the unhealthy node."

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,18 @@
+
+FAR remediates a node by rebooting an unhealthy node, and moving any remaining workloads from the node to other nodes, so they can be processed in other nodes and be isolated from the unhealthy node. The reboot is done by executing a fence agent for the unhealthy node, while evicting the workloads from this node is achieved by tainting the node and deleting the workloads.
+
+FAR uses a fence agent to fence a Kubernetes node. Generally, fencing is the process of taking unresponsive/unhealthy computers into a safe state and isolating the computer. Fence agent is a software code that uses a management interface to perform fencing, mostly power-based fencing which enables power-cycling, resetting, or turning off the computer. FAR includes some of the fence agents from th ...
```
"computers" sounds strange to me, no clue why to be honest. Maybe just because that term is seldom used in our area of work?
What about using "servers"?

razo7 replied:
Yes, it might be more appropriate

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,29 @@
+
+[Fence Agents Remediation](https://github.com/medik8s/fence-agents-remediation#readme) (FAR) is a Kubernetes operator that uses external tools to *fence* unhealthy nodes. The tools are a set of well-known [upstream fence agents (FA)](https://github.com/ClusterLabs/fence-agents), where each one of them can be used for different environments, to fence a node using a reboot action. Moreover, FAR can ...
+
+FAR is not only fencing a node when it becomes unhealthy, it also *remediates* the node from being unhealthy to healty again. It adds a taint to evict stateless pods, fence the node with a fence agent and fully remediate it with resource deletion to remove any remaining workloads (mostly statefull workloads) after reboot.
```
"it also *remediates* the node"
-> it also tries to remediate the node"
Everything we do is best effort. There is no guarantee that the reboot heals the node.

slintes replied:
statefull -> stateful

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,29 @@
+
+FAR is not only fencing a node when it becomes unhealthy, it also *remediates* the node from being unhealthy to healty again. It adds a taint to evict stateless pods, fence the node with a fence agent and fully remediate it with resource deletion to remove any remaining workloads (mostly statefull workloads) after reboot.
+
+FAR is recommended when a node becomes unhealthy, and we want to completely isolate the node from a cluster, since we can’t “trust” the unhealthy node, to prevent it from accessing the shared resources like [RWO volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes). The operator is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/oper ...
```
> FAR is recommended when a node becomes unhealthy, and we want to completely isolate the node from a cluster, since we can’t “trust” the unhealthy node, to prevent it from accessing the shared resources...

While this is true in general, this isn't specific to FAR, is it? This sounds like other remediators don't do this 🤔

razo7 replied:
>This sounds like other remediators don't do this

From the sentence beforehand I assume you mean that other operators act similar to FAR on that matter (when to use remediation).
Do you see any value of adding similar sentence to the [remediation page](https://www.medik8s.io/remediation/remediation/) if not here?

slintes replied:
I think we have it there already:

```
At it's core, remediation (aka. fencing) turns a question _Can our peer cause
data corruption?_ into an answer _**No!**_ by isolating it both from incoming
requests and persistent storage.
```

so it's just redundant here IMHO

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,29 @@
+
+### Advantages
+
+* Robustness - FAR can remediate an unhealthy node using a management interface (e.g., IPMI) while still keeping control plane connectivity.
```
> while still keeping control plane connectivity

what do you mean with this? 🤔

razo7 replied:
The term `control plane` might be confusing in our world, as it could hint to Kubernetes control plane node, but it wasn't my intention.
I meant that by using fence agent to reboot the node we receive an acknowledgment that the fence agent finished/failed/succeeded while still maintaining a connection to the node, and not losing it during/after the reboot process.

slintes replied:
> we receive an acknowledgment that the fence agent finished/failed/succeeded while still maintaining a connection to the node, and not losing it during/after the reboot process.

I think this is misleading. AFAIK we don't have a connection to the "node", and not to the control plane (which is k8s term and shouldn't be used for anything else than the actual k8s control plane IMHO). We connect to the "management  interface".
I guess what you are trying to point out here is: we have direct feedback if the fence action succeeded or not, right?
Maybe something like this is less confusing?

`Robustness - By using a management interface, FAR has direct feedback about the result of the fence action without having to use the Kubernetes API.`

razo7 replied:
>I guess what you are trying to point out here is: we have direct feedback if the fence action succeeded or not, right?

Correct

>We connect to the "management interface".

After your [last comment](https://github.com/medik8s/docs/pull/63#discussion_r1345695572) about the difference between `management interface` and  `traditional API`, I went with only using the API option. So unless you suggest using the other  `management interface` or explaining what this interface is earlier on this page, then I would prefer to use the API term.

>Robustness - By using a management interface, FAR has direct feedback about the result of the fence action without having to use the Kubernetes API.

Better, but I went with something else to keep a similar pattern/structure to the advantages bullets (begin the sentence with `FAR...`), and use `API` over the `management interface`.

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,29 @@
+### Advantages
+
+* Robustness - FAR can remediate an unhealthy node using a management interface (e.g., IPMI) while still keeping control plane connectivity.
+* Speed - FAR is fast as it adds the FAR NoExecute taint, and deletes workloads to accelerate workload rescheduling, in addition to rebooting the node using a fence agent.
```
> it adds the FAR NoExecute taint, and deletes workloads to accelerate workload rescheduling

SNR does the same.
The advantage over SNR with regards to speed is that FAR doesn't need to wait until it's "safe to assume the node rebooted", but instead FAR just knows directly when the reboot command was executed. Does that make sense?

razo7 replied:
Yes, thanks for the input

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,29 @@
+
+* Robustness - FAR can remediate an unhealthy node using a management interface (e.g., IPMI) while still keeping control plane connectivity.
+* Speed - FAR is fast as it adds the FAR NoExecute taint, and deletes workloads to accelerate workload rescheduling, in addition to rebooting the node using a fence agent.
+* Diversity - FAR includes several FAs from a whole known set of upstream fencing agents for bare metal servers, virtual machines, cloud platforms, etc.
```
FA isn't an "official" acronym, is it? I would prefer "fence agents"

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,29 @@
+* Speed - FAR is fast as it adds the FAR NoExecute taint, and deletes workloads to accelerate workload rescheduling, in addition to rebooting the node using a fence agent.
+* Diversity - FAR includes several FAs from a whole known set of upstream fencing agents for bare metal servers, virtual machines, cloud platforms, etc.
+* Adjustability - FAR allows to set up different parameters to access the management interface that remediates the node.
+* Compatibility with NHC - FAR works with NHC to provide high availability for nodes in an automatic manner. NHC detects the node’s health and triggers FAR to remediate the node until it becomes healthy again.
```
how is this an advantage, all our remediators do that 🙂

razo7 replied:
It is an advantage in general, and possibly a reminder in case it was not mentioned properly beforehand.
But I do see your point. How about `FAR works with NHC` -> `similar to other remediation operators by Medik8s, FAR can also work with NHC`

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,29 @@
+* Diversity - FAR includes several FAs from a whole known set of upstream fencing agents for bare metal servers, virtual machines, cloud platforms, etc.
+* Adjustability - FAR allows to set up different parameters to access the management interface that remediates the node.
+* Compatibility with NHC - FAR works with NHC to provide high availability for nodes in an automatic manner. NHC detects the node’s health and triggers FAR to remediate the node until it becomes healthy again.
+* Standalone - FAR can be used as a standalone solution to remediate nodes with an administrator's help. Moreover, FAR is not limited to NHC and could work with another detection mechanism.
```
What does "standalone" mean?
You can create the remediation CRs yourself, or use other detection mechanism, for all our remediators.
SNR can even remediate without any interaction 🤷🏼‍♂️

razo7 replied:
> You can create the remediation CRs yourself, or use other detection mechanism, for all our remediators.

Correct, I was thinking of expressing it so users won't think that FAR, similar to SNR and MDR, is limited to be used with NHC. I thought it would help to increase engagement in case users want to pre-automate the remediation manually or with another detecting mechanism to remediate a node.

It might better to use 'Semi-automation'.

>Semi-automation - FAR, and all the remediation operators from Medik8s, can semi-automate the node remediation once the node has been detected/selcted to be remediated. The selection can be with NHC, or another detection mechanism, or even an administrator's help.

razo7 replied:
I have decided to rephrase it again in a new paragraph prior to the advantages section

slintes commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -0,0 +1,17 @@
+
+## Fence Agents Remediation Configuration
+
+Fence Agents Remediation (FAR) uses a Custom Resource (CR) created by [Node Health Check Operator](https://github.com/medik8s/node-healthcheck-operator#readme) (NHC) based on a specific remediation template (*FenceAgentsRemediationTemplate*).
```
why "specific" remediation template?

razo7 replied:
I was trying to show the importance of using this template when using FAR with NHC. So that using any other template for FAR and NHC won't work. Is that kind of trivial that we should avoid of expressing or the word `specific` just does not fit here?

slintes replied:
IMHO using a template is trivial, it's how NHC works in gerneral

slintes commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -0,0 +1,17 @@
+
+Fence Agents Remediation (FAR) uses a Custom Resource (CR) created by [Node Health Check Operator](https://github.com/medik8s/node-healthcheck-operator#readme) (NHC) based on a specific remediation template (*FenceAgentsRemediationTemplate*).
+
+Since NHC comes with [Self Node Remediation](https://www.medik8s.io/remediation/self-node-remediation/self-node-remediation/) as the default remediation mechanism, it needs to be configured to use FAR. Thus a user must do the following steps:
```
no, SNR is installed, but not configured as default (anymore)

razo7 replied:
Aha, good to know :)

razo7 committed:
Addressing peer comments

Rephrase sentences, clarify sentences and terms

razo7 committed:
API call instead of management interface

In this context there is no much difference, and using traditional API or just API should be enough to explain how the fence agent is communicating.

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,20 @@
+nav_order: 1
+---
+
+## How does Fence Agents Remediation works?
```
works -> work

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,20 @@
+
+## How does Fence Agents Remediation works?
+
+The fence Agents Remediation (FAR) operator uses a traditional API to remediate unhealthy Kubernetes nodes.
```
fence -> Fence

slintes reviewed (changes requested):
left some comments inline.... I hope at last, somehow Goland seems to mess up the review 🙈

slintes commented on remediation/fence-agents-remediation/Limitations.md:
```diff
@@ -0,0 +1,12 @@
+---
```
please use lowercase filename, like the others are

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,41 @@
+The operator is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/operator/fence-agents-remediation), and can be installed manually as mentioned in [FAR documentation](https://github.com/medik8s/fence-agents-remediation#installation)
+
+Similar to the other [remediation operators from Medik8s](https://www.medik8s.io/remediation/remediation/#implementations),
+FAR can semi-automate the node remediation once the node has been detected/selected to be remediated.
```
- why semi-automate? It's fully automated, not? 🙂 
- FAR can automate -> FAR can be used to automate

razo7 replied:
>why semi-automate? It's fully automated, not? 

Yes, I think the term `automation` doesn't fit here.

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,41 @@
+FAR can semi-automate the node remediation once the node has been detected/selected to be remediated.
+The selection can be done with the [Node HealthCheck Operator (NHC)](https://github.com/medik8s/node-healthcheck-operator#readme) which detects the node’s health,
+another detection mechanism (or any other operator that supports an external remediation API), or even using an administrator’s help.
+To fully automate a remediation process we can combine NHC with FAR, and automatically increase the availability of workloads.
```
I think this sentence is redundant, NHC is mentioned above already

razo7 committed:
Remove automation, and summarize idesa

Rephrase sentences, clarify sentences and terms

slintes commented on remediation/remediation.md:
```diff
@@ -5,16 +5,18 @@ nav_order: 4
 ---
 
-# Putting Nodes into a Safe State (Remediation)
+## Putting Nodes into a Safe State (Remediation)
```
why did you change the heading level?

razo7 replied:
Due to markdown linter in the IDE, see _Multiple top-level headings in the same document_  ([MD025](https://github.com/DavidAnson/markdownlint/blob/v0.31.1/doc/md025.md)).

I can not make this change here (and create a top-level heading for new files of FAR) and only do that in a follow-up PR if it makes more sense.

clobrano replied:
I don't understand. The MD025 mentions too many top-level headings, but here there was only one

razo7 replied:
> The MD025 mentions too many top-level headings

It also mentions:
>This rule is triggered when a top-level heading is in use (the first line of the file is an h1 heading), and more than one h1 heading is in use in the document:

razo7 replied:
> but here there was only one

No, IIUC there is one for the title coming from https://github.com/medik8s/docs/pull/63/files#diff-dfff4e442e5bc623a5e3bbd60058c511d647599ede49b9c2a876eb4b546b5d61R3, and the rest are also h1 headings (with 1 #)

clobrano replied:
> there is one for the title coming from #63 (files)

This is not rendered as a Title in the page though, it's the name of the file in the sidebar.

> the rest are also h1 headings (with 1 #)

I see only "Putting Nodes into a Safe State (Remediation)" as H1 heading [in the main (gh-pages) branch](https://raw.githubusercontent.com/medik8s/docs/gh-pages/remediation/remediation.md) :thinking:

razo7 replied:
>it's the name of the file in the sidebar.

Correct and the name of the page as [title](https://github.com/medik8s/docs/blob/gh-pages/remediation/remediation.md?plain=1#L3) .
![image](https://github.com/medik8s/docs/assets/19728420/9085608c-1459-4b70-b1b8-6710483ae320)

clobrano replied:
Yes, my point being maybe the linter is wrong 😁

slintes replied:
> I can not make this change here

Why not? If the linter is the problem, disable that rule 🤷🏼‍♂️ The page title is taken from the title attribute, not the heading, so that rule doesn't make sense here.

razo7 replied:
>I can not make this change here

I basically can, even though I wrote differently. The word `not` was misplaced here 🙅🏻.
Anyway. I will not follow this remark/rule and ignore it here (and on the other new pages)

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,20 @@
+nav_order: 1
+---
+
+## How does Fence Agents Remediation work?
```
why not top level heading, as we do on every other page?

razo7 replied:
See my comment above

slintes commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,20 @@
+The remediation includes rebooting the unhealthy node using a fence agent and evicting workloads from the unhealthy node by resource deletion.
+
+FAR uses a fence agent to fence a Kubernetes node. Generally, fencing is the process of taking unresponsive/unhealthy servers into a safe state and isolating the server.
+Fence agent is a software code that uses an Application Programming Interface (API) call to perform fencing, mostly power-based fencing which enables power-cycling, resetting, or turning off the server.
```
software code? :thinking:

razo7 replied:
Yes, in the end, it is a Python script. But we could mention it differently.
How about just `code` or `script`?
I thought of highlighting that since it  is different than SNR which uses a hardware device (but I might be wrong here because I remember that there is a software reboot command with SNR, right?)
Furthermore, I wanted to be more specific and elaborate on what does fence agent term really means.

slintes replied:
`application`?

razo7 replied:
SGTM

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,37 @@
+
+### Advantages
+
+* Robustness - FAR has direct feedback from the traditional API call (e.g., IPMI) about the result of the fence action without using the Kubernetes API.
```
typo ` e.g.,`

razo7 replied:
Can you elaborate? Do I miss a space or `e.g.,` was not used correctly?
I looked at https://www.aje.com/arc/editing-tip-using-eg-and-ie/ and it looks ok,

slintes replied:
sorry, I was under the impression that there is no need for a comma

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,37 @@
+
+* Robustness - FAR has direct feedback from the traditional API call (e.g., IPMI) about the result of the fence action without using the Kubernetes API.
+* Speed - FAR is rapid since it can reboot a node and receive an acknowledgment from the API call while others might need to wait a safe time till they can expect the node to be rebooted.
+Similarly to some other operators, FAR adds a NoExecute taint and deletes workloads to accelerate workload rescheduling, in addition to rebooting the node.
```
I think this isn't related to the "speed" advantage

razo7 replied:
Isn't accelerating the isolation of the node from its resources won't help FAR to be faster?  Or you are not referring to this as an advantage for FAR (since SNR does it as well)?

slintes replied:
the latter

razo7 replied:
I will move this sentence to a different place since I find it valuable information.

slintes commented on remediation/fence-agents-remediation/limitations.md:
```diff
@@ -0,0 +1,12 @@
+## Limitations
+
+1. FAR is limited to running fence agents with only one [action](https://github.com/ClusterLabs/fence-agents/blob/main/doc/FenceAgentAPI.md#agent-operations-and-return-values), reboot.
+2. FAR remediation time can be long when the FAR pod resides in the unhealthy node since the FAR pod needs to be rescheduled on a new node, and that could already be tainted (with NoExecute), thus it would take more time.
```
I do not understand, how does the taint slow down rescheduling of the FAR pod?

razo7 replied:
>how does the taint slow down rescheduling of the FAR pod?

It is not the taint by itself. The remediation time is just slower when the FAR pod is installed on the unhealthy node because then it would need to be rescheduled on a different node. The pod would be evicted by remediation if the taint hasn't evicted it already.

slintes replied:
> needs to be rescheduled on a new node, and that could already be tainted

- this is confusing, it sounds like the new node is tainted
- what taint actually?
- this is too detailed IMHO
- maybe just stop after `on a new node`?

> The pod would be evicted by remediation

which remediation?

razo7 replied:
>which remediation?

I meant that the resource deletion. As it would accelerate the rescheduling of the FAR pod if the taint didn't evict the pod already. 

>this is confusing, it sounds like the new node is tainted

I see your point. Thanks for mentioning that.

>what taint actually?

ATM there is only one taint in FAR context. FAR NoExecute taint.

>this is too detailed IMHO
maybe just stop after on a new node?

SGTM

slintes reviewed (changes requested):
Looking at fence-agents-remediation and how-it-works, there is much duplicated content. Would it make sense to simplify this? Either keep the former much more simple, or get rid of the latter? :thinking:

clobrano commented on remediation/remediation.md:
```diff
@@ -5,16 +5,18 @@ nav_order: 4
-## The Problem
+### The Problem
 
 A common approach to for providing high availability for Kubernetes is to run
```
Typo: remove "to"

Also, not sure about "providing" later in the paragraph, it does not fit with previous verb:

"to run ... and start ..."

clobrano replied:
Nevermind, this is not part of the PR. Let's consider it a reminder for a future fix :)

razo7 replied:
You can open up a PR for this manner :)

clobrano commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,37 @@
+It adds a taint to evict stateless pods, fence the node with a fence agent, and after reboot it completes the remediation with
+resource deletion to remove any remaining workloads (mostly stateful workloads).
+
+Similar to the other [remediation operators from Medik8s](https://www.medik8s.io/remediation/remediation/#implementations), FAR can be used to remediate a node that has been detected/selected to be remediated.
```
> FAR can be used to remediate a node that has been detected/selected to be remediated

It sounds a bit complicated to read to me. The core is that it can work both standalone or in conjunction with another tool, in a more automatic fashion, right?

razo7 replied:
Correct

clobrano replied:
So I think would be easier to understand if saying that directly, like

"Far, similarly ..., can be used standalone, manually creating a CR, or together with another operator that automatically detects/select the target node and create the CR via a template"

or something like that

razo7 replied:
>the core is that it can work both standalone or in conjunction with another tool,

The paragraph is to introduce the option of using another tool with FAR. I eventually added the "where to find FAR in the community" senetence, since I thought it could align here.
>So I think would be easier to understand if saying that directly

In the next sentence, we mention how FAR can work with NHC, another detecting mechanism (not necessarily an operator) or alone but with the use of an admin. I fixed to sentence tagged sentence.

clobrano replied:
I think it is still unclear that this selection/detection process is automatic.

> The selection can be done with the Node HealthCheck Operator

while it might be clearer if re-phrased somehow like

> The selection can be made by NHC, which automatically detects... and creates the CR for...

Although, "selection" does not sound right to me. Maybe just "detected unhealthy".

razo7 replied:
>>while it might be clearer if re-phrased somehow like

>The selection can be made by NHC, which automatically detects... and creates the CR for...

But this way the detection seems automatic and not the far cr creation. Both are important, but only the far CR creation that can happen automatically is important for FAR POV.
The paragraph is not about NHC working automatically rather FAR can work automatically using another tool/operator.
ATM the creation is mentioned to happen automatically, but maybe it was not stressed out properly being written in parenthesis. Is that what bugs you?

razo7 replied:
>Although, "selection" does not sound right to me. Maybe just "detected unhealthy".

 remediates a node ... for the unhealthy detected node.
The detection can be done ...

razo7 replied:
>selection/detection process is automatic.

The detection by NHC is not mentioned to be automatic but the far CR creation does (see below).
>which detects the node’s health (and automatically creates far CR)

I think the latter is more important and related to FAR. Do you still think it is important? 

>Although, "selection" does not sound right to me. Maybe just "detected unhealthy".

I changed it a little bit.

clobrano commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,37 @@
+another detection mechanism (or any other operator that supports an external remediation API), or even manually using an administrator’s help.
+Moreover, FAR is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/operator/fence-agents-remediation), and can be installed manually as mentioned in [FAR documentation](https://github.com/medik8s/fence-agents-remediation#installation).
+
+As opposed to other remediation systems, FAR requires credentials for running the API call (e.g., Intelligent Platform Management Interface ([IPMI](https://en.wikipedia.org/wiki/Intelligent_Platform_Management_Interface)), etc.)
```
I see "in contrast" more appropriate than "as opposed". Or just differently.

razo7 committed:
Ignore MD025 markdownlint rule

Multiple top-level headings in the same document [MD025](https://github.com/DavidAnson/markdownlint/blob/v0.31.1/doc/md025.md) rule complains here about the title field as h1 heading (but it is only the page name), thus it doesn't fit here.

razo7 committed:
Simplify FAR limitation and rephrase worload rescheduling

Simplify FAR limitation, fence agent is considered as an application, and rephrase worload rescheduling

razo7 commented on remediation/remediation.md:
```diff
@@ -52,8 +55,9 @@ single cluster, without creating conflicts.
 * [(wip) metal3](https://github.com/metal3-io/cluster-api-provider-metal3/pull/157) - a mechanism designed for bare metal clusters with a functioning Metal3 API
 * [machine-deletion-remediation](https://github.com/medik8s/machine-deletion-remediation) - a mechanism designed for any cluster with a functioning Machine API
-* [(planned) direct]() - a mechanism designed around an existing set of [upstream fencing agents](https://github.com/ClusterLabs/fence-agents) for environments with a traditional API end-point (eg. IPMI) for power cycling cluster nodes
+* [fence-agents-remediation](https://github.com/medik8s/fence-agents-remediation) - a mechanism designed around an existing set of [upstream fencing agents](https://github.com/ClusterLabs/fence-agents) for environments with a traditional API end-point (eg. IPMI) for power-cycling the cluster nodes
```
Do we want to prioritize FAR and rank it higher in this list and in [navigation bar](https://github.com/medik8s/docs/pull/63/files#diff-292f5d801f6deb757dcd72b8b27a9cb338cf46664452fdb4b91016ba4aece3d8R4)?

razo7 replied:
We agreed internally, that there is no meaning for this order here. It is unordered list

clobrano commented on remediation/remediation.md:
```diff
@@ -52,8 +55,9 @@ single cluster, without creating conflicts.
 * [(wip) metal3](https://github.com/metal3-io/cluster-api-provider-metal3/pull/157) - a mechanism designed for bare metal clusters with a functioning Metal3 API
 * [machine-deletion-remediation](https://github.com/medik8s/machine-deletion-remediation) - a mechanism designed for any cluster with a functioning Machine API
-* [(planned) direct]() - a mechanism designed around an existing set of [upstream fencing agents](https://github.com/ClusterLabs/fence-agents) for environments with a traditional API end-point (eg. IPMI) for power cycling cluster nodes
+* [fence-agents-remediation](https://github.com/medik8s/fence-agents-remediation) - a mechanism designed around an existing set of [upstream fencing agents](https://github.com/ClusterLabs/fence-agents) for environments with a traditional API end-point (eg. IPMI) for power-cycling the cluster nodes
```
Typo `s/eg./e.g./`

clobrano commented on remediation/fence-agents-remediation/how-it-works.md:
```diff
@@ -0,0 +1,21 @@
+---
+
+<!-- markdownlint-disable-next-line MD025 -->
+# How does Fence Agents Remediation work?
```
This section seems to repeat the same information of fence-agents-remediation.md

razo7 replied:
It indeed includes similar information as in `fence-agents-remediation.md` (mostly in the first paragraph), but it adds a little bit more information about the concept of fencing and fence agents in the second paragraph.

slintes replied:
I agree with Carlo, this site doesn't add any value... that very little bit which wasn't mentioned yet can be moved to the parent site.

slintes replied:
(or the other way around...)

clobrano reviewed (changes requested):
I left some comments

razo7 committed:
Simplify FAR similarity sentence and mentioning automation

clobrano commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -22,12 +22,13 @@ It adds a taint to evict stateless pods ([NoExecute](https://kubernetes.io/docs/
-The selection can be done with the [Node HealthCheck Operator (NHC)](https://github.com/medik8s/node-healthcheck-operator#readme) which detects the node’s health,
-another detection mechanism (or any other operator that supports an external remediation API), or even manually using an administrator’s help.
-Moreover, FAR is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/operator/fence-agents-remediation), and can be installed manually as mentioned in [FAR documentation](https://github.com/medik8s/fence-agents-remediation#installation).
+FAR, similarly to the other [remediation operators from Medik8s](https://www.medik8s.io/remediation/remediation/#implementations), remediates an unhealthy node by creating its own [custome resource (CR)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
```
sorry, typo :sweat_smile: 
s/custome/custom

clobrano replied:
moreover, it seems that FAR creates the CR by itself written this way

razo7 replied:
>it seems that FAR creates the CR by itself written this way

Yes, it was a little bit confusing.

clobrano commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -22,12 +22,13 @@ It adds a taint to evict stateless pods ([NoExecute](https://kubernetes.io/docs/
+FAR, similarly to the other [remediation operators from Medik8s](https://www.medik8s.io/remediation/remediation/#implementations), remediates an unhealthy node by creating its own [custome resource (CR)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+for that node.
+The node can be detected as unhealthy using the [Node HealthCheck Operator (NHC)](https://github.com/medik8s/node-healthcheck-operator#readme) which checks the node’s health
+(and automatically creates far CR when it is needed), another detection mechanism (or any other operator that supports an external remediation API), or even manually using an administrator’s help.
```
nit: no need to put the sentence under brackets (IMHO)

razo7 replied:
Which one or both of them?

clobrano replied:
I was referring to the first

razo7 committed:
Clarify that FAR doesn't create far CR

From the sentence one can understand that FAR creates the CR by itself

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,39 @@
+
+FAR is not only fencing a node when it becomes unhealthy, it also tries to *remediate* the node from being unhealthy to healthy again.
+It adds a taint to evict stateless pods ([NoExecute](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions)), fence the node with a fence agent,
+and after reboot it completes the remediation with resource deletion to remove any remaining workloads (mostly stateful workloads).
```
I would skip "resource deletion", it's too verbose IMHO. What about "it completes the remediation by deleting remaining workloads (mostly stateful workloads)."

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,39 @@
+FAR, similarly to the other [remediation operators from Medik8s](https://www.medik8s.io/remediation/remediation/#implementations), remediates an unhealthy node when its own
+[custom resource (CR)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) is created for that node.
+The node can be detected as unhealthy using the [Node HealthCheck Operator (NHC)](https://github.com/medik8s/node-healthcheck-operator#readme) which checks the node’s health
+and automatically creates far CR when it is needed, another detection mechanism (or any other operator that supports an external remediation API), or even manually using an administrator’s help.
```
nit: creates the FAR CR

slintes replied:
I would add another "or" before "another detection" in order to improve readability for the very long sentence. Or split into multiple sentences.

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,39 @@
+
+FAR, similarly to the other [remediation operators from Medik8s](https://www.medik8s.io/remediation/remediation/#implementations), remediates an unhealthy node when its own
+[custom resource (CR)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) is created for that node.
+The node can be detected as unhealthy using the [Node HealthCheck Operator (NHC)](https://github.com/medik8s/node-healthcheck-operator#readme) which checks the node’s health
```
"can be detected as unhealthy" and "checks the node’s health" is duplicate.
Maybe just ", which will create FAR CRs automatically."

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,39 @@
+[custom resource (CR)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) is created for that node.
+The node can be detected as unhealthy using the [Node HealthCheck Operator (NHC)](https://github.com/medik8s/node-healthcheck-operator#readme) which checks the node’s health
+and automatically creates far CR when it is needed, another detection mechanism (or any other operator that supports an external remediation API), or even manually using an administrator’s help.
+Moreover, FAR is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/operator/fence-agents-remediation), and it can even be installed manually as mentioned in [FAR documentation](https://github.com/medik8s/fence-agents-remediation#installation).
```
- unneeded transition word "Moreover"
- "and" should be "or" IMHO
- "even" can be skipped

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,39 @@
+and automatically creates far CR when it is needed, another detection mechanism (or any other operator that supports an external remediation API), or even manually using an administrator’s help.
+Moreover, FAR is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/operator/fence-agents-remediation), and it can even be installed manually as mentioned in [FAR documentation](https://github.com/medik8s/fence-agents-remediation#installation).
+
+In contrast to other remediation systems, FAR requires credentials for running the API call (e.g., Intelligent Platform Management Interface ([IPMI](https://en.wikipedia.org/wiki/Intelligent_Platform_Management_Interface)), etc.)
```
IPMI isn't an API call... ?
Maybe "for executing the API call to e.g. the IPMI"

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,39 @@
+and automatically creates far CR when it is needed, another detection mechanism (or any other operator that supports an external remediation API), or even manually using an administrator’s help.
+Moreover, FAR is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/operator/fence-agents-remediation), and it can even be installed manually as mentioned in [FAR documentation](https://github.com/medik8s/fence-agents-remediation#installation).
+
+In contrast to other remediation systems, FAR requires credentials for running the API call (e.g., Intelligent Platform Management Interface ([IPMI](https://en.wikipedia.org/wiki/Intelligent_Platform_Management_Interface)), etc.)
```
This sentence sounds like "hey this is a disadvantage but otherwise we are great!".
I think this is a wrong attitude. People using IPMI etc. very likely know that such API calls need authorization/authentication.
I would skip the complete sentence here, it's a configuration detail IMHO.

razo7 replied:
>People using IPMI etc. very likely know that such API calls need authorization/authentication.

Ok, and the unfamiliar users will eventually learn about that from searching online on IPMI (or a different management interface).

slintes commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -0,0 +1,18 @@
+<!-- markdownlint-disable-next-line MD025 -->
+# Fence Agents Remediation Configuration
+
+Fence Agents Remediation (FAR) uses a Custom Resource (CR) created by [Node Health Check Operator](https://github.com/medik8s/node-healthcheck-operator#readme) (NHC) based on a FAR's remediation template (*FenceAgentsRemediationTemplate*).
```
Elsewhere you always mention that FAR CRs can be created manually. And here you just mention NHC.
What about using 2 sections here, one for FAR CRs, and one about the templates with NHC?

razo7 committed:
Rewrite FAR parent page, how it works and configuration pages

Move FAR explanation from parent page to how it works, and split configuration to standalone and FAR with NHC. Some grammar improvements and new links

slintes commented on remediation/fence-agents-remediation/fence-agents-remediation.md:
```diff
@@ -0,0 +1,26 @@
+The remediation includes rebooting the unhealthy node using a fence agent and then evicting workloads from the unhealthy node.
+
+FAR is available in the Kubernetes community, [OperatorHub.io](https://operatorhub.io/operator/fence-agents-remediation), or it can be installed manually as mentioned in [FAR documentation](https://github.com/medik8s/fence-agents-remediation#installation).
+Below are his main advantages.
```
s/his/its
Also, IMHO this is redundant. You can see that the next section, just one line below, is about advantages. WDYT?

razo7 replied:
Yes, I was not sure how much this line was going to be helpful for the user. I will remove it.

slintes commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -0,0 +1,32 @@
+# Fence Agents Remediation Configuration
+
+Fence Agents Remediation (FAR) operator watches *FenceAgentsRemediation* (or *far*) Custom Resource ([CR](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)).
+FAR can run alongside another operator (e.g., Node Health Check Operator [(NHC)](https://github.com/medik8s/node-healthcheck-operator#readme)) or another mechnaisam that trigger FAR by automatically creating far CRs using *FenceAgentsRemediationTemplate* (or *fartemplate*) CR. FAR can also be run as standalone, when far CR is generated manually using an administrator’s help.
```
what do you mean with "FAR can run alongside another operator"?
What about something like: Those CRs can be created either manually by cluster administrators, or by other operators (e.g. ...) using FenceAgentsRemediationTemplates ...

slintes commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -0,0 +1,32 @@
+* Create fartemplate CR
+* Create NHC CR to use fartemplate CR accordingly (see [NHC documentation](https://github.com/medik8s/node-healthcheck-operator/blob/main/docs/configuration.md#remediationtemplate)).
+
+<!-- Fence Agents Remediation (FAR) uses a Custom Resource (CR) created by [Node Health Check Operator](https://github.com/medik8s/node-healthcheck-operator#readme) (NHC) based on a FAR's remediation template (*FenceAgentsRemediationTemplate*). -->
```
what's this for?

razo7 replied:
Redundant, I have forgotten to remove it

slintes commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -0,0 +1,32 @@
+
+## FAR Standalone
+
+Administrator should choose the fence agent based on the cluster platform, and know its parameters for running FAR properly.
```
this is not specific to standalone, you need to the same for creating a template

razo7 replied:
I will move it after to the subsections.

slintes commented on remediation/fence-agents-remediation/limitations.md:
```diff
@@ -0,0 +1,13 @@
+<!-- markdownlint-disable-next-line MD025 -->
+# Limitations
+
+1. FAR is limited to running fence agents with only one [action](https://github.com/ClusterLabs/fence-agents/blob/main/doc/FenceAgentAPI.md#agent-operations-and-return-values), reboot.
```
"one" can be understood wrong, as in "one" vs multiple. Maybe this is a bit clearer: Currently FAR supports the "reboot" fence agent action only.

razo7 replied:
I haven't notice that. Also, I liked the use of `support` vs `limited to`.

slintes commented on remediation/fence-agents-remediation/limitations.md:
```diff
@@ -0,0 +1,13 @@
+# Limitations
+
+1. FAR is limited to running fence agents with only one [action](https://github.com/ClusterLabs/fence-agents/blob/main/doc/FenceAgentAPI.md#agent-operations-and-return-values), reboot.
+2. FAR remediation time can be long when the FAR pod resides in the unhealthy node since the FAR pod needs to be rescheduled on a different node.
```
Maybe also add "Currently" here, as it will be resolved soon

slintes replied:
Also, I think this more a known issue than a limitation, is it? :thinking:

slintes reviewed (changes requested):
good pogress :+1:
some comments inline

razo7 committed:
Move an issue to a limitation and reorder a general note on using FAR

Addressing comments and improve docs clarity

slintes commented on remediation/fence-agents-remediation/configuration.md:
```diff
@@ -10,7 +10,7 @@ nav_order: 2
 
 Fence Agents Remediation (FAR) operator watches *FenceAgentsRemediation* (or *far*) Custom Resource ([CR](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)).
-FAR can run alongside another operator (e.g., Node Health Check Operator [(NHC)](https://github.com/medik8s/node-healthcheck-operator#readme)) or another mechnaisam that trigger FAR by automatically creating far CRs using *FenceAgentsRemediationTemplate* (or *fartemplate*) CR. FAR can also be run as standalone, when far CR is generated manually using an administrator’s help.
+Those CRs can be created either manually by cluster administrators (and run as standalone), or by other operators (e.g., Node Health Check Operator [(NHC)](https://github.com/medik8s/node-healthcheck-operator#readme)) or different mechnaisams that trigger FAR by automatically creating far CRs using *FenceAgentsRemediationTemplate* (or *fartemplate*) CR.
```
> (and run as standalone)

- too verbose IMHO
- also wrong, because "CRs can be created and run as standalone" doesn't make sense

> or different mechnaisams

- typo
- what about "or similar tools"?

razo7 replied:
>what about "or similar tools"?

👍🏻 

>too verbose IMHO
also wrong, because "CRs can be created and run as standalone" doesn't make sense

I was thinking of adding this word (`standalone`) before showing it in the subsection title, so the user will understand what is this subsection.

razo7 committed:
Add recommendation setence for FAR with NHC

razo7 committed:
Fix tyo

s/simialr/similar
